File: extension/src/sync/dormantSyncSystem.ts
import { SYNC_ACTIVATION } from "./activation";

export interface AuthenticatedOperationIngress {
  /** The only synchronization ingress; implementations delegate to OperationKernel.ingest. */
  ingest(signedEnvelope: Uint8Array): Promise<string>;
}
export interface SyncActivationMode { readonly productionActivated: boolean; readonly testArtifact: boolean }

export class DormantSyncSystem {
  #started = false;
  constructor(private readonly ingress: AuthenticatedOperationIngress, private readonly mode: SyncActivationMode = SYNC_ACTIVATION) {}
  startTestArtifact(): void {
    if (!this.mode.testArtifact || this.mode.productionActivated) throw new Error("Dormant synchronization is unavailable in production");
    this.#started = true;
  }
  start(): void {
    if (!this.mode.testArtifact && !this.mode.productionActivated) throw new Error("Dormant synchronization is unavailable in production");
    this.#started = true;
  }
  async ingestFromReplica(signedEnvelope: Uint8Array): Promise<string> {
    if (!this.#started) throw new Error("Dormant synchronization is unavailable in production");
    return await this.ingress.ingest(signedEnvelope.slice());
  }
  productionMigrationCutoverAllowed(): false { return false; }
}

File: app/src/main/java/com/pomo/sync/DormantSyncSystem.kt
package com.pomo.sync

internal fun interface AuthenticatedOperationIngress {
    /** The only synchronization ingress; implementations delegate to OperationKernel.ingest. */
    fun ingest(signedEnvelope: ByteArray): String
}

internal class DormantSyncSystem(
    private val ingress: AuthenticatedOperationIngress,
    private val mode: SyncActivationMode = syncActivationMode,
) {
    private var testSystemStarted = false

    fun startTestArtifact() {
        check(mode.testArtifact && !mode.productionActivated)
        testSystemStarted = true
    }

    fun start() {
        check(mode.testArtifact || mode.productionActivated) {
            "Dormant synchronization is unavailable in production"
        }
        testSystemStarted = true
    }

    fun ingestFromReplica(signedEnvelope: ByteArray): String {
        check(testSystemStarted) { "Dormant synchronization is unavailable in production" }
        return ingress.ingest(signedEnvelope.copyOf())
    }

    fun productionMigrationCutoverAllowed(): Boolean = false
}
